libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshController.kt

```diff
@@ -0,0 +1,65 @@
+package com.mapbox.navigation.core.routerefresh
+
+import android.util.Log
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.navigation.core.trip.session.TripSession
+import com.mapbox.navigation.utils.timer.MapboxTimer
+import java.util.concurrent.TimeUnit
+import kotlinx.coroutines.Job
+
+/**
+ * This class is responsible for refreshing the current direction route's traffic at a
+ * specified [intervalSeconds]. This does not support alternative routes.
+ *
+ * If the route is route is successfully refreshed, this class will update the [TripSession.route]
+ *
+ * [start] and [stop] are attached to the application lifecycle. Observing routes that
+ * can be refreshed are handled by this class. Calling [start] will restart the refresh timer.
+ */
+internal class RouteRefreshController(
+    private var accessToken: String,
+    private val tripSession: TripSession
+) {
+    private val routerRefreshTimer = MapboxTimer()
+    private val routeRefreshRetrofit = RouteRefreshRetrofit()
+    private val routeRefreshApi = RouteRefreshApi(routeRefreshRetrofit)
+
+    var intervalSeconds: Long = TimeUnit.MILLISECONDS.toSeconds(routerRefreshTimer.restartAfterMillis)
+        set(value) {
+            routerRefreshTimer.restartAfterMillis = TimeUnit.SECONDS.toMillis(value)
+            field = value
+        }
+
+    fun start(): Job {
+        stop()
+        return routerRefreshTimer.startTimer {
+            if (routeRefreshApi.supportsRefresh(tripSession.route)) {
+                routeRefreshApi.refreshRoute(
+                    accessToken,
+                    tripSession.route,
+                    tripSession.getRouteProgress(),
+                    routeRefreshCallback)
+            }
+        }
+    }
+
+    fun stop() {
+        routerRefreshTimer.stopJobs()
+    }
+
+    private val routeRefreshCallback = object : RouteRefreshCallback {
+
+        override fun onRefresh(directionsRoute: DirectionsRoute) {
+            Log.i("RouteRefresh", "Successful refresh")
+            tripSession.route = directionsRoute
```

Comment:
Should we update the `MapboxDirectionsSession#routes[0]` too 🤔? Actually, updating the `MapboxDirectionsSession#routes` will end up updating `MapboxTripSession#route` 👀 https://github.com/mapbox/mapbox-navigation-android/blob/1656229a9bbe80b541314c1522b67b686b8e3143/libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt#L440-L448

Reply:
Also noticed that we're updating the whole route in NN https://github.com/mapbox/mapbox-navigation-android/blob/1656229a9bbe80b541314c1522b67b686b8e3143/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt#L45 instead of only the annotations as we do in the legacy 👀 [`RouteHandler`](https://github.com/mapbox/mapbox-navigation-android/blob/master/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/RouteHandler.kt). Do you know if this has a performance impact? Probably yes because it's not the same to serialize the whole route rather than only updating the annotations of the current route.

Reply:
if the traffic profile changes, NN should know about it right?

Reply:
looking into update the annotations only now, i did try before but that was early on.

considered performance impacts and considered it minor since it's per minutes (opposed to per seconds)

Reply:
Another note is that legacy doesn't only update the navigator annotations. It also updates the entire route
https://github.com/mapbox/mapbox-navigation-android/blob/master/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/RouteRefresherCallback.kt#L17

Reply:
Nah, that's not correct 👀 https://github.com/mapbox/mapbox-navigation-android/blob/cd4d864c4ab4d40e83b42564feb6ceeb80b2b54a/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/RouteHandler.kt#L18-L26 if it's not a `NEW_ROUTE` https://github.com/mapbox/mapbox-navigation-android/blob/cd4d864c4ab4d40e83b42564feb6ceeb80b2b54a/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/RouteRefresherCallback.kt#L17 we're only updating the annotations.

Reply:
Yeah discussed offline, now see how fresh_route causes the condition in RouteHandler

Reply:
When attempting to only update annotations in NN, it didn't update the the traffic on the map.